### ISSUE TITLE
4.14.0 breaks builds because of change in complexity rule (#8535)

### ISSUE THREAD

**Issue by epoberezkin** [eslint/eslint#9765]
**Tell us about your environment**

* **ESLint Version:**
4.14.0
* **Node Version:**
4-9
* **npm Version:**

**What parser (default, Babel-ESLint, etc.) are you using?**
Default

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

The failing builds are cause by the “fix” of complexity algorithm in #8535.
Should it not be based on some option until the major version change?

```bash
eslint
```

**What did you expect to happen?**
pass

**What actually happened? Please include the actual, raw output from ESLint.**
fail with the error from complexity rule being higher than the threshold.

**Comment by platinumazure** [comment]
Hi @epoberezkin, thanks for the issue.

Being a lint tool, we are aware of how bug fixes can be more disruptive than usual to our users due to the critical role ESLint plays in a lot of projects' CI setups. At the same time, we do want to be able to fix bugs without having to label many bug fixes as breaking changes.

We've come up with a specific [semantic versioning policy](https://github.com/eslint/eslint/blob/master/README.md#semantic-versioning-policy) to reflect this. Basically, we agree to publish all bug fixes which increase warnings for a particular rule as semver-minor rather than semver-patch; and in exchange, we guarantee that patch upgrades will never break end users' builds, but minor upgrades might for the purpose of fixing bugs. We also recommend pinning ESLint version dependencies to the minor version number instead of major (i.e., use `~` instead of `^`).

Regarding the change to `complexity`: The difference between `&&` and `||` in that rule has been long-standing, but it is a bug that we basically inherited from JSHint. As noted in the discussion in the related issue, `&&` and `||` are equivalent in terms of complexity: `a && b === !(!a || !b)`.

I would recommend holding off on upgrading ESLint (if needed) until you can fix any issues newly raised by the rule.

**Comment by epoberezkin** [comment]
Understood, thank you. Please feel free to close it then.
